### PR TITLE
zos: build in ASCII code page and introduce zoslib

### PR DESCRIPTION
**Pull request by zsw007**
This PR is for building libuv on z/OS systems. The PR make changes to build libuv in ASCII code page, and introduce a dependency on [ZOSLIB](https://github.com/ibmruntimes/zoslib) to enable this transition.

[ZOSLIB](https://github.com/ibmruntimes/zoslib) is a z/OS C/C++ library which implements additional POSIX APIs that are not available in the LE C Runtime Library, and provides API for EBCDIC <-> ASCII conversion. ZOSLIB can be installed separately on the system, and then linked to libuv using the `-DZOSLIB_DIR` option introduced by this PR.

### CODE REVIEW

**Comment by cjihrig** [comment]
cc @libuv/zos

**Comment by richardlau** [comment]
Changes look okay to me but I'm holding off from officially approving as we need to make a few build changes so we can test this in the CI. I've raised https://github.com/nodejs/build/issues/2491 to cover that and hope to get that done over the next week or two.

**Comment by richardlau** [comment]
I've been working on a copy of the z/OS libuv job with modifications to build this PR in https://ci.nodejs.org/view/libuv/job/libuv-test-commit-zos-cmake-richardlau/

The modified job can now build `zoslib` (the new dependency) and libuv (with this PR) but there are some test failures: https://ci.nodejs.org/view/libuv/job/libuv-test-commit-zos-cmake-richardlau/47

@zsw007 If you can confirm those test failures are expected to be addressed in follow up PR's I'd be okay approving this and coordinating the switchover for the Jenkins job configuration. The two failures for the statically compiled test (`uv_run_tests_a`) we have discussed in https://github.com/nodejs/build/issues/2491#issuecomment-750636512 but we have several more failures in the dynamically list test suite (`uv_run_tests`)
e.g. https://ci.nodejs.org/view/libuv/job/libuv-test-commit-zos-cmake-richardlau/47/BUILDTYPE=Debug,nodes=zos-s390x/console
```
not ok 15 - closed_fd_events
# exit code 11
# Output from process `closed_fd_events`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
...
not ok 141 - ipc_closed_handle
# exit code 11
# Output from process `ipc_closed_handle`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
not ok 142 - ipc_heavy_traffic_deadlock_bug
# exit code 11
# Output from process `ipc_heavy_traffic_deadlock_bug`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
not ok 143 - ipc_listen_after_write
# exit code 11
# Output from process `ipc_listen_after_write`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
not ok 144 - ipc_listen_before_write
# exit code 11
# Output from process `ipc_listen_before_write`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
not ok 145 - ipc_send_recv_pipe
# exit code 11
# Output from process `ipc_send_recv_pipe`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
ok 146 - ipc_send_recv_pipe_inprocess
not ok 147 - ipc_send_recv_tcp
# exit code 11
# Output from process `ipc_send_recv_tcp`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
ok 148 - ipc_send_recv_tcp_inprocess
not ok 149 - ipc_send_zero
# exit code 11
# Output from process `ipc_send_zero`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
not ok 150 - ipc_tcp_connection
# exit code 11
# Output from process `ipc_tcp_connection`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
not ok 151 - kill
# exit code 11
# Output from process `kill`:
# CEE3204S The system detected a protection exception (System Completion Code=0C4).
#          From compile unit S0W1:/u/iojs/build/workspace/libuv-test-commit-zos-cmake-richardlau/BUILDTYPE/Debug/nodes/zos-s390x/src/strscpy.c at entry point uv__strscpy at statement 29 at compile unit offset +000000001C0480B6 at entry offset 
#          +0000000000000076 at address 000000001C0480B6.
...
```

**Review by richardlau — COMMENTED** [review]
@zsw007 Please could you PR some words into appropriate places in [`README.md`](https://github.com/libuv/libuv/blob/v1.x/README.md) that mention `zoslib` as a dependency for z/OS and having to pass `ZOSLIB_DIR` to `cmake`?

**Comment by zsw007** [comment]
> @zsw007 If you can confirm those test failures are expected to be addressed in follow up PR's I'd be okay approving this and coordinating the switchover for the Jenkins job configuration. The two failures for the statically compiled test (`uv_run_tests_a`) we have discussed in [nodejs/build#2491 (comment)](https://github.com/nodejs/build/issues/2491#issuecomment-750636512) but we have several more failures in the dynamically list test suite (`uv_run_tests`)

I have included the fixes for the two failures in `uv_run_tests_a`, but the failures in `uv_run_tests` are unexpected. I will take a look at them.

**Comment by zsw007** [comment]
@richardlau We made an update to ZOSLIB to address the failures in the dynamically linked tests. Both static and dynamic tests should be passing now when using ZOSLIB commit 5920b41. Could you please try running the job again to verify?

**Comment by richardlau** [comment]
https://ci.nodejs.org/view/libuv/job/libuv-test-commit-zos-cmake-richardlau/71/ passed 🎉.

**Review by richardlau — APPROVED** [review]

**Comment by richardlau** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2113/
Resume-CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2114/

CI looks good (one instance of https://github.com/libuv/libuv/issues/2307 and the `libuv-test-commit-zos-cmake` job is expected to fail with this PR until it is updated as per https://github.com/nodejs/build/issues/2491). We'll need to coordinate landing this PR and updating https://ci.nodejs.org/view/libuv/job/libuv-test-commit-zos-cmake to avoid broken builds. 

@cjihrig what's the current outlook on https://github.com/libuv/libuv/issues/3086? I'd like not to break the CI while a release is in progress. FWIW I don't think this PR needs to make 1.41.0 (but equally no objections if it does).

**Comment by cjihrig** [comment]
The release is currently waiting on https://github.com/libuv/libuv/pull/3064 (cc: @vtjnash). After that lands, I'll also need to run the CI and Node.js integration CI before cutting the release.

**Pull request merged**
